mono/metadata/mono-perfcounters.c

```diff
@@ -1283,61 +1317,65 @@ custom_get_value_address (SharedCounter *scounter, SharedInstance* sinst)
 }
 
 static void*
-custom_get_impl (SharedCategory *cat, MonoString *counter, MonoString* instance, int *type, MonoError *error)
+custom_get_impl (SharedCategory *cat, mono_unwrapped_string_t counter,
+		mono_unwrapped_string_t instance, int *type, MonoError *error)
 {
-	SharedCounter *scounter;
-	SharedInstance* inst;
-	char *name;
-
-	error_init (error);
-	scounter = find_custom_counter (cat, counter);
+	SharedCounter * const scounter = find_custom_counter (cat, counter);
 	if (!scounter)
 		return NULL;
-	name = mono_string_to_utf8_checked (counter, error);
+
+	char * const name = mono_unwrapped_string_to_utf8 (counter, error);
 	return_val_if_nok (error, NULL);
+
 	*type = simple_type_to_type [scounter->type];
-	inst = custom_get_instance (cat, scounter, name);
+
+	SharedInstance* const inst = custom_get_instance (cat, scounter, name);
 	g_free (name);
 	if (!inst)
 		return NULL;
 	return custom_vtable (scounter, inst, (char *)custom_get_value_address (scounter, inst));
 }
 
 static const CategoryDesc*
-find_category (MonoString *category)
+find_category (mono_unwrapped_string_t category)
 {
-	int i;
-	for (i = 0; i < NUM_CATEGORIES; ++i) {
-		if (mono_string_compare_ascii (category, predef_categories [i].name) == 0)
-			return &predef_categories [i];
+	const CategoryDesc* desc = predef_categories;
+	for (int i = 0; i < NUM_CATEGORIES; ++i, ++desc) {
+		if (perfcounter_utf16_equal_ascii (category.chars, category.length, desc->name, desc->name_length))
+			return desc;
 	}
 	return NULL;
 }
 
 void*
-mono_perfcounter_get_impl (MonoString* category, MonoString* counter, MonoString* instance,
-		MonoString* machine, int *type, MonoBoolean *custom)
+ves_icall_System_Diagnostics_PerformanceCounter_GetImpl (
+		MonoStringHandle category_handle, MonoStringHandle counter_handle, MonoStringHandle instance_handle,
+		MonoStringHandle machine, int *type, MonoBoolean *custom, MonoError *error)
 {
-	MonoError error;
-	const CategoryDesc *cdesc;
+	error_init (error);
+
 	void *result = NULL;
+	gchar *c_instance = NULL;
+	mono_unwrapped_string_t category = mono_unwrap_string_handle (category_handle);
```

Comment:
Use the `fixed` keyword in managed to pin the object in memory. It makes the code much more readable and allows you to manipulate directly `gunichar2*` memory in native.

Reply:
New-ish approach to all this work?